common/quotas/global/algorithm/requestweighted.go

```diff
@@ -480,10 +514,20 @@ func (a *impl) getWeightsLocked(key Limit, snap configSnapshot) (weights map[Ide
 		return nil, 0, met
 	}
 
+	// zeros anywhere here should not be possible - they are prevented from being inserted,
+	// and anything simply "losing weight" will only become "rather low", not zero,
+	// before enough passes have occurred to garbage collect it.
+	//
+	// specifically, 1rps -> 0rps takes over 1,000 halving cycles, and a single 1rps event
+	// during that will immediately re-set it above 0.5 and will need 1,000+ more cycles.
+	//
+	// if gc period / weight amount is set extreme enough this is "possible",
+	// but we are unlikely to ever cause it.
 	for id := range ir {
-		// scale by the total.
-		// this also ensures all values are between 0 and 1 (inclusive)
-		weights[id] = weights[id] / total
+		// normalize by the total.
+		// this also ensures all values are between 0 and 1 (inclusive),
+		// though zero should be impossible.
+		weights[id] = shared.SanityLogFloat(0, weights[id]/total, 1, "normalized weight", a.logger)
```

Comment:
another important fix: if `total` was 0 (e.g. say an aggregator just started receiving requests, and it received only a 0 from a previously-used-but-idle limit), this would return a NaN to the host that submitted the 0.

this is now prevented by not recording zeros at all, and the sanity-log exists to validate that assumption.

I don't believe this is related to the "sustained 0rps" issue, because it's not recording a NaN anywhere, and should correct itself as soon as non-zero traffic arrived (so: a single update round after any request).  but it could be responsible for brief periods of limiting.